trust_stores_observatory/store_fetcher/java_fetcher.py

```diff
@@ -0,0 +1,115 @@
+            fh2.flush()
+            result = subprocess.run(['keytool', '-list', '-v', '-keystore', fh2.name], stdout=subprocess.PIPE, input=b'')
+            result = result.stdout.decode('utf-8')
+    except Exception as inst:
+      raise ValueError('Could not fetch file')
+    else:
+      parsed_root_records = self._get_root_records(result)
+      trusted_certificates = RootRecordsValidator.validate_with_repository(cert_repo, hashes.SHA256(), parsed_root_records)  
+  
+    return TrustStore(PlatformEnum.ORACLE_JAVA, version_num, url, datetime.utcnow().date(), 
+                      trusted_certificates)
+    
+  @staticmethod
+  def _get_root_records(raw_file: str) -> List[Tuple[str,bytes]]:
+    # asterisks seperate each certificate record
+    result = re.split(r'\*+', raw_file)
+    root_records = []
+
+    for item in result:
+      owner_index = item.find('Owner')
+      if owner_index != -1:
+        n = item[owner_index:]
+        m = n.split('\n')
+        owner_name = m[0]
+        potential_subject = re.search('((?:CN=|OU=)([^,]+),?)',owner_name)
+        if potential_subject is not None:
+          subject_name = potential_subject.group(2)
+        fingerprint = None
+        for inner in m:
+          if 'SHA256' not in inner:
+            continue
+          fingerprint_hex = re.search('(?<=SHA256:)(.+)', inner)
+          if fingerprint_hex is not None:
+            fingerprint_hex = fingerprint_hex.group(0).replace(':', '').strip()
+            fingerprint = bytes(bytearray.fromhex(fingerprint_hex))
+            break
+
+        root_records.append((subject_name, fingerprint))
+    return root_records  
+
+  @classmethod
+  def _get_latest_package_url(cls) -> Tuple[str,str]:
+
+    with urlopen(cls._BASE_URL + cls._DOWNLOADS_INDEX) as response:
+      page_content = response.read()
+    parsed_page = BeautifulSoup(page_content, 'html.parser')
+
+    href = parsed_page.find('img',alt='Download JRE').parent
+    latest_link = href.get('href')
+
+    with urlopen(cls._BASE_URL+latest_link) as download_page:
+      download_content = download_page.read()
+    parsed = BeautifulSoup(download_content, 'html.parser')
+
+    #table is populated dynamically so we need to use regex to extract link
+    download_array = re.findall("downloads\[.+\]", str(parsed))
+    tar_files = [p for p in download_array if p.endswith("tar.gz']")]
+    potential_target_file = tar_files[-1]
+    potential_target_file = re.search("\['files'\]\['(.+tar\.gz)'", potential_target_file)
+    if potential_target_file is not None:
+      link = potential_target_file.group(1)
+      potential_filepath = re.search('.+'+link, str(parsed))
+      potential_filepath = potential_filepath.group(0)
+      version = re.search("jre-(\w+\.\w+\.\w+)", potential_filepath).group(1)
+      filepath = re.search("(?<=\"filepath\":)(.+)", potential_filepath)
+      filepath = filepath.group(0).replace('"', '')
+      return filepath, version
```

Comment:
There's a lot of regular expressions being used here, but regexps should really be last resort (they have a very high chance of introducing non-trivial bugs due to an unexpected match and they are hard to read). 
Can you instead use BeautifulSoup to walk the page's DOM and find the element you are looking for (without a regexp)?

Reply:
Hi!

First, I want to say thank you for reviewing my code! I understand you're busy and I really, really appreciate you taking time to review!

Regarding regex, the reason I'm using regex here and not BeautifulSoup is because the downloads table is  generated dynamically via inline JavaScript. I originally tried to use BeautifulSoup to find the element but was unsuccessful as the element wasn't on the page,  which is why I started to use regex instead. An alternative would be to use a headless browser to parse the page but then this would add extra overhead to your project. 

I could get rid of the extra regex and keep the lines that find the files: 

` download_array = re.findall("downloads\[.+\]", str(parsed))`
`tar_files = [p for p in download_array if p.endswith("tar.gz']")]`

this would still require string manipulation to some extent to get the file path and version number. 

What do you think??

Reply:
Yeah it makes sense why you can’t use BeautifulSoup. As for the regexps, I haven’t looked at the content of the page but any way you can just use `parsed.split(‘downloads’, 1)` and similar string manipulation functions, instead of a full blown regexp? If not then regexps are fine.

Reply:
Thanks for being patient with me as I work through this, and also for suggesting useful tools and ideas :)) I took your suggestion and used `find` and `rfind` to extract the necessary information and it worked really well!! Plus, it's much easier to understand than regexp. 

I also used pyjks and it's really neat! The only parsing I had to do manually was extracting the subject name, and the error I mentioned above ( subject_name not being in repo ) was gone when I refreshed the stores 🎉 

Let me know what you think!